#include <iostream>
#include <memory>
#include <string>
#include <thread>
#include <fstream>
#include <sstream>
#include <vector>
#include <filesystem>
#include <algorithm> 
#include <semaphore>
#include <cstring> // for memcmp

#include <grpcpp/grpcpp.h>
#include <sw/redis++/redis++.h>
#include <nlohmann/json.hpp>

#include "judge.grpc.pb.h" 
#include "sandbox.h"
#include "etcd_registry.h"
// å¼•å…¥ Internal å¤´æ–‡ä»¶è¯»å–é…ç½®
#include "sandbox_internal.h" 

// å£°æ˜ Config åŠ è½½å‡½æ•°
void LoadConfig(const std::string& path);

using json = nlohmann::json;
using grpc::Server;
using grpc::ServerBuilder;
using grpc::ServerContext;
using grpc::Status;
namespace fs = std::filesystem;

// å…¨å±€ Redis è¿æ¥
std::shared_ptr<sw::redis::Redis> g_redis = nullptr;

// å…¨å±€å¹¶å‘æ§åˆ¶ä¿¡å·é‡
std::unique_ptr<std::counting_semaphore<>> g_task_sem = nullptr;

// ---------------------------------------------------------
// åˆ¤é¢˜æ ¸å¿ƒï¼šæµå¼æ¯”å¯¹ (Stream Comparator)
// ---------------------------------------------------------

// è¿”å› true è¡¨ç¤º ACï¼Œfalse è¡¨ç¤º WA
bool check_output(const std::string& user_out_path, const std::string& std_ans_path) {
    if (!fs::exists(std_ans_path)) {
        std::cerr << "[æ¯”å¯¹] æ ‡å‡†ç­”æ¡ˆç¼ºå¤±: " << std_ans_path << std::endl;
        return false;
    }
    if (!fs::exists(user_out_path)) {
        std::cerr << "[æ¯”å¯¹] ç”¨æˆ·è¾“å‡ºç¼ºå¤±: " << user_out_path << std::endl;
        return false;
    }

    std::ifstream f_user(user_out_path);
    std::ifstream f_std(std_ans_path);

    // è‡ªåŠ¨è·³è¿‡ whitespace è¿›è¡Œ token-based æ¯”è¾ƒ
    std::string s_u, s_s;
    std::cout << "[æ¯”å¯¹] æ­£åœ¨æ¯”è¾ƒ ç”¨æˆ·è¾“å‡º: " << user_out_path << " vs æ ‡å‡†ç­”æ¡ˆ: " << std_ans_path << std::endl;
    while (true) {
        bool b_u = (bool)(f_user >> s_u);
        bool b_s = (bool)(f_std >> s_s);

        // å¦‚æœä¸€ä¸ªè¯»å®Œäº†ï¼Œå¦ä¸€ä¸ªæ²¡è¯»å®Œ -> WA
        if (b_u != b_s) {
            std::cout << "[æ¯”å¯¹] é•¿åº¦ä¸ä¸€è‡´! ç”¨æˆ·è¾“å‡ºç»“æŸ=" << !b_u << ", æ ‡å‡†ç­”æ¡ˆç»“æŸ=" << !b_s << std::endl;
            // Debug: print contents
            std::ifstream f_dbg(user_out_path);
            std::stringstream buffer;
            buffer << f_dbg.rdbuf();
            std::cout << "[æ¯”å¯¹] ç”¨æˆ·å†…å®¹ DUMP: [" << buffer.str() << "]" << std::endl;
            return false;
        }

        // å¦‚æœéƒ½è¯»å®Œäº† -> AC
        if (!b_u) return true;

        // å¦‚æœå†…å®¹ä¸ä¸€æ · -> WA
        if (s_u != s_s) {
            std::cout << "[æ¯”å¯¹] å†…å®¹ä¸åŒ¹é…! ç”¨æˆ·: [" << s_u << "] vs æ ‡å‡†ç­”æ¡ˆ: [" << s_s << "]" << std::endl;
            return false;
        }
    }
}

// ---------------------------------------------------------
// åå°å·¥ä½œçº¿ç¨‹
// ---------------------------------------------------------
void process_task(deep_oj::TaskRequest task) {
    std::string job_id = task.job_id();
    std::cout << "[Worker] å¼€å§‹åå°å¤„ç†: " << job_id << std::endl;

    // 1. åˆå§‹åŒ–æ²™ç®±
    deep_oj::Sandbox sandbox(deep_oj::g_runner_config.workspace_root);
    json result_json;
    result_json["job_id"] = job_id;

    // æ³¨ï¼šæºç å†™å…¥ç”± sandbox.Compile() å†…éƒ¨å®Œæˆï¼Œæ— éœ€åœ¨æ­¤é‡å¤å†™å…¥

    // 2. ç¼–è¯‘
    deep_oj::CompileResult cres = sandbox.Compile(job_id, task.code());

    if (!cres.success) {
        result_json["status"] = "Compile Error";
        result_json["error"] = cres.error_message;
        goto REPORT;
    }

    // 3. è¿è¡Œ & åˆ¤é¢˜
    {
        // å‡è®¾é¢˜ç›® ID å¯¹åº”çš„æµ‹è¯•æ•°æ®ç›®å½• (ç”Ÿäº§ç¯å¢ƒåº”ä»é…ç½®æˆ– task è¯»å–)
        // [TODO] ç”Ÿäº§ç¯å¢ƒéœ€å¾ªç¯æµ‹è¯•æ‰€æœ‰ç”¨ä¾‹
        std::string std_in = "data/1.in";   // ä½ çš„æµ‹è¯•è¾“å…¥
        std::string std_out = "data/1.out"; // ä½ çš„æ ‡å‡†ç­”æ¡ˆ
        
        // è·å–å·¥ä½œç›®å½•å¹¶æ„å»ºç»å¯¹è·¯å¾„
        fs::path work_dir = fs::path(cres.exe_path).parent_path();
        fs::path output_path = work_dir / "output.txt";
        fs::path error_path = work_dir / "error.txt";
        
        deep_oj::RunResult rres = sandbox.Run(
            cres.exe_path, 
            std_in,                 // stdin 
            output_path.string(),   // stdout (ç»å¯¹è·¯å¾„)
            error_path.string(),    // stderr (ç»å¯¹è·¯å¾„)
            task.time_limit(), 
            task.memory_limit()
        );

        result_json["time_used"] = rres.time_used;
        result_json["memory_used"] = rres.memory_used;

        if (rres.status == deep_oj::SandboxStatus::OK) {
            // æ²™ç®±è¿è¡ŒæˆåŠŸï¼Œç°åœ¨è¿›è¡Œç­”æ¡ˆæ¯”å¯¹
            bool is_ac = check_output(output_path.string(), std_out);
            
            if (is_ac) {
                result_json["status"] = "Accepted";
                result_json["judge_result"] = 1; 
            } else {
                result_json["status"] = "Wrong Answer";
                result_json["judge_result"] = 2; 
            }
        } else {
            // å¼‚å¸¸çŠ¶æ€æ˜ å°„
            if (rres.status == deep_oj::SandboxStatus::TIME_LIMIT_EXCEEDED) {
                result_json["status"] = "Time Limit Exceeded";
            }
            else if (rres.status == deep_oj::SandboxStatus::MEMORY_LIMIT_EXCEEDED) {
                result_json["status"] = "Memory Limit Exceeded";
            }
            else if (rres.status == deep_oj::SandboxStatus::OUTPUT_LIMIT_EXCEEDED) {
                result_json["status"] = "Output Limit Exceeded";
            }
            else {
                // Runtime Error å¤„ç†é€»è¾‘
                result_json["status"] = "Runtime Error";
                result_json["exit_code"] = rres.exit_code; 
                
                // 1. è¯»å–æ²™ç®±å†…çš„ stderr
                std::ifstream err_fs(error_path);
                if (err_fs.good()) {
                    std::stringstream buffer;
                    buffer << err_fs.rdbuf();
                    std::string err_msg = buffer.str();
                    
                    if (!err_msg.empty()) {
                        // å°† stderr å†…å®¹å›å¡«åˆ° JSONï¼Œå‰ç«¯å¯ä»¥å¼¹çª—æ˜¾ç¤º
                        result_json["error_message"] = "[Sandbox Stderr]: " + err_msg;
                        std::cout << "[Worker] Stderr: " << err_msg << std::endl;
                    } else {
                        result_json["error_message"] = "Process exited abnormally with no output.";
                    }
                }
                
                // 2. å¦‚æœæ˜¯ Signal å¯¼è‡´çš„ (å¦‚ SegFault)ï¼ŒSandbox.cpp å·²ç»å¡«äº† error_message
                if (!rres.error_message.empty()) {
                     // ä¼˜å…ˆå±•ç¤º Sandbox æ•è·çš„ä¿¡å·æè¿°
                     result_json["error_message"] = rres.error_message;
                }
            }
        }
    }

REPORT:
    // 4. ç»“æœä¸ŠæŠ¥
    if (g_redis) {
        try {
            std::string res_str = result_json.dump();
            g_redis->set("result:" + job_id, res_str);
            g_redis->expire("result:" + job_id, 3600);
            
            // åªæœ‰ AC æ‰å†™ Cache
            if (!task.cache_key().empty() && result_json["status"] == "Accepted") {
                g_redis->set(task.cache_key(), res_str);
                g_redis->expire(task.cache_key(), 86400); 
            }
            
            // é…åˆ Scheduler çš„ LREMï¼Œå¿…é¡»å‘é€å®Œæ•´çš„ä»»åŠ¡æ•°æ®
            // ä¹‹å‰åªå‘ job_id ä¼šå¯¼è‡´ Scheduler æ— æ³•åœ¨ Listener ä¸­åŒ¹é…åˆ°é˜Ÿåˆ—é‡Œçš„å…ƒç´ 
            g_redis->publish("job_done", job_id);
            
            std::cout << "[Worker] âœ… å®Œæˆ: " << job_id << " (" << result_json["status"] << ")" << std::endl;
        } catch (const std::exception& e) {
            std::cerr << "âŒ [Worker] ç»“æœä¸ŠæŠ¥å¤±è´¥ (" << job_id << "): " << e.what() << std::endl;
        }
    }
    
    // æ¸…ç†ä¸´æ—¶æ–‡ä»¶
    sandbox.Cleanup(job_id);
}

// å…¨å±€ä»»åŠ¡å»é‡ (é˜²æ­¢ Double Dispatch å¯¼è‡´çš„ç«æ€)
std::mutex g_job_mutex;
std::set<std::string> g_running_jobs;

class WorkerServiceImpl final : public deep_oj::JudgeService::Service {
    Status ExecuteTask(ServerContext* context, const deep_oj::TaskRequest* request,
                  deep_oj::TaskResponse* response) override {
        
        std::string job_id = request->job_id();
        std::cout << "[Worker] æ”¶åˆ°è°ƒåº¦è¯·æ±‚ (ID: " << job_id << ")" << std::endl;

        // 1. å¹‚ç­‰æ€§æ£€æŸ¥
        {
            std::lock_guard<std::mutex> lock(g_job_mutex);
            if (g_running_jobs.count(job_id)) {
                std::cout << "[Worker] âš ï¸ Ignored duplicate request: " << job_id << std::endl;
                return Status::OK; // è§†ä¸ºæˆåŠŸï¼Œä¸é‡å¤æ‰§è¡Œ
            }
            g_running_jobs.insert(job_id);
        }

        if (!g_task_sem->try_acquire()) {
            std::cout << "[Worker] âš ï¸ High Load - æ‹’ç»è¯·æ±‚: " << job_id << std::endl;
            // é‡Šæ”¾ Job ID
            {
                std::lock_guard<std::mutex> lock(g_job_mutex);
                g_running_jobs.erase(job_id);
            }
            return Status(grpc::RESOURCE_EXHAUSTED, "Worker is busy");
        }

        std::thread([req_copy = *request]() mutable {
            std::string jid = req_copy.job_id();
            
            // RAII Cleanup Guard
            struct ResourceGuard {
                std::string id;
                ResourceGuard(std::string i) : id(i) {}
                ~ResourceGuard() { 
                    g_task_sem->release();
                    std::lock_guard<std::mutex> lock(g_job_mutex);
                    g_running_jobs.erase(id);
                }
            } guard(jid);
            
            process_task(std::move(req_copy));
        }).detach();

        return Status::OK;
    }
};

int main(int argc, char** argv) {
    // 1. åŠ è½½é…ç½® (å¿…é¡»åœ¨æ‰€æœ‰é€»è¾‘ä¹‹å‰)
    // 1. åŠ è½½é…ç½® (ä¼˜å…ˆä» config/worker.yaml åŠ è½½)
    if (std::filesystem::exists("config/worker.yaml")) {
        LoadConfig("config/worker.yaml");
    } else {
        LoadConfig("config.yaml");
    }

    // 2. Redis è¿æ¥
    try {
        const char* env_redis = std::getenv("REDIS_HOST");
        std::string redis_host = env_redis ? env_redis : "127.0.0.1";
        g_redis = std::make_shared<sw::redis::Redis>("tcp://" + redis_host + ":6379");
        std::cout << "[Worker] Redis è¿æ¥æˆåŠŸ" << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "âŒ [Fatal] Redis è¿æ¥å¤±è´¥" << std::endl;
        return 1;
    }

    // 3. åˆå§‹åŒ–å…¨å±€ä¿¡å·é‡
    int pool_size = deep_oj::g_runner_config.pool_size;
    if (pool_size <= 0) pool_size = 4; // å…œåº•
    g_task_sem = std::make_unique<std::counting_semaphore<>>(pool_size);
    std::cout << "[Worker] ğŸ”¥ å¹¶å‘æ¨¡å‹å·²åˆå§‹åŒ–: Max Threads = " << pool_size << std::endl;

    // 4. å¯åŠ¨ gRPC Server
    int port = deep_oj::g_runner_config.server_port;
    if (port <= 0) port = 50051;
    std::string server_address("0.0.0.0:" + std::to_string(port));
    
    WorkerServiceImpl service;

    ServerBuilder builder;
    builder.AddListeningPort(server_address, grpc::InsecureServerCredentials());
    builder.RegisterService(&service);
    
    std::unique_ptr<Server> server(builder.BuildAndStart());
    std::cout << "ğŸš€ [Worker] å¯åŠ¨ç›‘å¬: " << server_address << std::endl;

    // 5. [å¯é€‰] Etcd æœåŠ¡å‘ç°æ³¨å†Œ
    // è®¾ç½® ETCD_ENDPOINTS ç¯å¢ƒå˜é‡ä»¥å¯ç”¨ (ä¾‹å¦‚: "127.0.0.1:2379")
    // ä¸è®¾ç½®åˆ™è·³è¿‡ï¼Œä¿æŒå‘åå…¼å®¹
    std::unique_ptr<deep_oj::EtcdRegistry> etcd;
    const char* etcd_endpoints = std::getenv("ETCD_ENDPOINTS");
    if (etcd_endpoints && strlen(etcd_endpoints) > 0) {
        try {
            const char* env_worker_id = std::getenv("WORKER_ID");
            std::string worker_id = env_worker_id ? env_worker_id : ("worker_" + std::to_string(port));
            
            // å¯¹å¤–åœ°å€: ä¼˜å…ˆä½¿ç”¨ WORKER_ADDRï¼Œå¦åˆ™ç”¨ server_address
            const char* env_addr = std::getenv("WORKER_ADDR");
            std::string advertise_addr = env_addr ? env_addr : server_address;

            etcd = std::make_unique<deep_oj::EtcdRegistry>(etcd_endpoints, 10);
            if (etcd->Register(worker_id, advertise_addr)) {
                std::cout << "[Worker] âœ… Etcd æ³¨å†ŒæˆåŠŸ: " << worker_id << " -> " << advertise_addr << std::endl;
            } else {
                std::cerr << "[Worker] âš ï¸ Etcd æ³¨å†Œå¤±è´¥ï¼Œç»§ç»­è¿è¡Œ (æ— æœåŠ¡å‘ç°)" << std::endl;
                etcd.reset();
            }
        } catch (const std::exception& e) {
            std::cerr << "[Worker] âš ï¸ Etcd åˆå§‹åŒ–å¼‚å¸¸: " << e.what() << "ï¼Œç»§ç»­è¿è¡Œ" << std::endl;
            etcd.reset();
        }
    }

    server->Wait();

    // 6. ä¼˜é›…å…³é—­: EtcdRegistry ææ„å‡½æ•°è‡ªåŠ¨ Deregister
    return 0;
}